File: apps/web/services/applied-learning/appliedLearning.ts
import { getAPIUrl } from '@services/config/config'
import { RequestBodyWithAuthHeader, errorHandling, secureFetch } from '@services/utils/ts/requests'

export type AppliedLearningEntry = {
  id: number
  entry_uuid: string
  user_id: number
  org_id: number
  course_uuid: string
  activity_uuid: string
  activity_name: string
  module_name: string
  planned_application: string
  previous_application: string
  measurable_change: string
  evidence_notes: string
  application_status: 'planned' | 'applied' | 'measured'
  created_at: string
  updated_at: string
}

export type AppliedLearningCapstone = {
  id: number
  capstone_uuid: string
  user_id: number
  org_id: number
  title: string
  challenge: string
  what_i_applied: string
  measurable_impact: string
  lessons_learned: string
  next_steps: string
  selected_entry_uuids: string[]
  status: 'draft' | 'ready' | 'submitted'
  created_at: string
  updated_at: string
}

export type ReflectionPayload = {
  org_id: number
  course_uuid: string
  activity_uuid: string
  activity_name?: string
  module_name?: string
  planned_application: string
  previous_application?: string
  measurable_change?: string
  evidence_notes?: string
  application_status?: 'planned' | 'applied' | 'measured'
}

function normalizeCourseUuid(value: string) {
  return value.startsWith('course_') ? value : `course_${value}`
}

function normalizeActivityUuid(value: string) {
  return value.startsWith('activity_') ? value : `activity_${value}`
}

export async function getAppliedLearningReflection(activityUuid: string, token?: string) {
  const normalizedUuid = normalizeActivityUuid(activityUuid)
  const res = await secureFetch(
    `${getAPIUrl()}applied-learning/reflection/${encodeURIComponent(normalizedUuid)}`,
    RequestBodyWithAuthHeader('GET', null, null, token)
  )
  return errorHandling(res) as Promise<AppliedLearningEntry | null>
}

export async function saveAppliedLearningReflection(payload: ReflectionPayload, token?: string) {
  const normalizedPayload: ReflectionPayload = {
    ...payload,
    course_uuid: normalizeCourseUuid(payload.course_uuid),
    activity_uuid: normalizeActivityUuid(payload.activity_uuid),
  }
  const res = await secureFetch(
    `${getAPIUrl()}applied-learning/reflection`,
    RequestBodyWithAuthHeader('POST', normalizedPayload, null, token)
  )
  return errorHandling(res) as Promise<AppliedLearningEntry>
}

export async function getMyAppliedLearning(orgId?: number, token?: string) {
  const qs = orgId ? `?org_id=${orgId}` : ''
  const res = await secureFetch(
    `${getAPIUrl()}applied-learning/me${qs}`,
    RequestBodyWithAuthHeader('GET', null, null, token)
  )
  return errorHandling(res) as Promise<AppliedLearningEntry[]>
}

export async function getAppliedLearningSummary(orgId?: number, token?: string) {
  const qs = orgId ? `?org_id=${orgId}` : ''
  const res = await secureFetch(
    `${getAPIUrl()}applied-learning/me/summary${qs}`,
    RequestBodyWithAuthHeader('GET', null, null, token)
  )
  return errorHandling(res) as Promise<{ entries: number; applied: number; measured: number; courses: number }>
}

export async function getMyCapstones(orgId?: number, token?: string) {
  const qs = orgId ? `?org_id=${orgId}` : ''
  const res = await secureFetch(
    `${getAPIUrl()}applied-learning/capstones/me${qs}`,
    RequestBodyWithAuthHeader('GET', null, null, token)
  )
  return errorHandling(res) as Promise<AppliedLearningCapstone[]>
}

export async function saveCapstone(payload: Partial<AppliedLearningCapstone> & { org_id: number; title: string }, token?: string) {
  const res = await secureFetch(
    `${getAPIUrl()}applied-learning/capstones`,
    RequestBodyWithAuthHeader('POST', payload, null, token)
  )
  return errorHandling(res) as Promise<AppliedLearningCapstone>
}

File: apps/api/src/db/applied_learning.py
from datetime import datetime, timezone
from typing import Optional

from sqlmodel import SQLModel, Field, Column, JSON


def utc_now_iso() -> str:
    return datetime.now(timezone.utc).replace(tzinfo=None).isoformat()


class AppliedLearningEntry(SQLModel, table=True):
    """A learner's durable record of applying a lesson in real work."""

    __tablename__ = "applied_learning_entry"

    id: Optional[int] = Field(default=None, primary_key=True)
    entry_uuid: str = Field(index=True, unique=True)
    user_id: int = Field(index=True)
    org_id: int = Field(index=True)
    course_uuid: str = Field(index=True)
    activity_uuid: str = Field(index=True)
    activity_name: str = ""
    module_name: str = ""
    planned_application: str = ""
    previous_application: str = ""
    measurable_change: str = ""
    evidence_notes: str = ""
    application_status: str = Field(default="planned", index=True)
    created_at: str = Field(default_factory=utc_now_iso)
    updated_at: str = Field(default_factory=utc_now_iso)


class AppliedLearningEntryRead(SQLModel):
    id: int
    entry_uuid: str
    user_id: int
    org_id: int
    course_uuid: str
    activity_uuid: str
    activity_name: str
    module_name: str
    planned_application: str
    previous_application: str
    measurable_change: str
    evidence_notes: str
    application_status: str
    created_at: str
    updated_at: str


class AppliedLearningCapstone(SQLModel, table=True):
    """A capstone draft assembled from selected applied-learning entries."""

    __tablename__ = "applied_learning_capstone"

    id: Optional[int] = Field(default=None, primary_key=True)
    capstone_uuid: str = Field(index=True, unique=True)
    user_id: int = Field(index=True)
    org_id: int = Field(index=True)
    title: str
    challenge: str = ""
    what_i_applied: str = ""
    measurable_impact: str = ""
    lessons_learned: str = ""
    next_steps: str = ""
    selected_entry_uuids: list[str] = Field(default_factory=list, sa_column=Column(JSON))
    status: str = Field(default="draft", index=True)
    created_at: str = Field(default_factory=utc_now_iso)
    updated_at: str = Field(default_factory=utc_now_iso)


class AppliedLearningCapstoneRead(SQLModel):
    id: int
    capstone_uuid: str
    user_id: int
    org_id: int
    title: str
    challenge: str
    what_i_applied: str
    measurable_impact: str
    lessons_learned: str
    next_steps: str
    selected_entry_uuids: list[str]
    status: str
    created_at: str
    updated_at: str
